### DISCUSSION TITLE
Explicit initialization without namespace

### DISCUSSION THREAD

**Discussion by totpero** [dotnet/csharplang#1214] [General] [answered]
Scenario 1:

I have my own declaration of SelectListItem and if i try this:
```
var my = new List<Base.Models.Metadate.SelectListItem>()
{
	new SelectListItem(),
       ....
}
```
I get ambiguous error because in my MVC app is another class with same name (https://msdn.microsoft.com/en-us/library/system.web.mvc.selectlistitem(v=vs.118).aspx) 

I'm forced to use the entire namespace path explicit like this:

```
var my = new List<Base.Models.Metadate.SelectListItem>()
{
	new Base.Models.Metadate.SelectListItem()
}
```

Is not obvious and sufficient if i have `List<Base.Models.Metadate.SelectListItem>()` ?

Scenario 2:

If i have this explicit variable declaration also i'm forced to use all namespace when i initialize this variable: 

`Base.Models.Metadate.SelectListItem my = new Base.Models.Metadate.SelectListItem();`

Why not using like this:

`Base.Models.Metadate.SelectListItem my = new SelectListItem();`

**Comment by Vlad-Stryapko** [comment]
You can make the existing case a bit simpler by namespace aliases.

> using Base = Base.Models.Metadate;  
> ...
> var item = new Base.SelectListItem();

Also note the usage of var instead of explicit type, which simplifies things for your second scenario.

**Comment by vladd** [comment]
Well, but what should happen in the following setup:

    var c = new NS1.Container<NS2.T2, NS3.T3>() // items are of type NS4.T4
    {
        new V() // is it NS1.V, NS2.V, NS3.V or NS4.V?
    };

? I'd better prefer target-typed `new`:

    var my = new List<Base.Models.Metadate.SelectListItem>()
    {
        new () // selects the item type automatically
    };

**Comment by totpero** [comment]
@vladd YES, this should be much better, elegant and simple way.

**Comment by CyrusNajmabadi** [comment]
> Is not obvious and sufficient if i have List<Base.Models.Metadate.SelectListItem>()

No, it is not.  You may be referencing another type with that same name from some other namespace.  Code like this already may exist today.  We can't break the meaning of it.

**Comment by totpero** [comment]
If I have the same name from same other namespace then I can get the ambiguous error.
Why would I name my assembly/namespaces like Microsoft?

**Comment by jnm2** [comment]
Even better is what ReSharper suggests and I'm hoping Roslyn soon suggests out of the box (https://github.com/dotnet/roslyn/issues/23373):

```cs
using SelectListItem = Base.Models.Metadate.SelectListItem;
// ...
var my = new List<SelectListItem>
{
    new SelectListItem()
}
```

With the championed target-typed `new` expression (https://github.com/dotnet/csharplang/issues/100) which I'm super excited about:

```cs
var my = new List<Base.Models.Metadate.SelectListItem>
{
    new()
}
```

**Comment by totpero** [comment, marked as answer]
In c# 9 this will be possible using [Target-typed new expressions ](https://github.com/dotnet/csharplang/blob/master/proposals/target-typed-new.md) 👍
